### PR TITLE
[ios] Map view can now be optionally configured to make pitch adjustments w…

### PR DESCRIPTION
**Pull request by d-prukop**
Map view can now be optionally configured to make pitch adjustments without affecting altitude. 

This addresses part of #9808

Pitch changes using MGLAltitudeForZoomLevel and MGLZoomLevelForAltitude from the MGLGeometry library were animating in a strange fashion when using the new proposed camera animation system (located here: https://github.com/mapbox/mapbox-ios-map-camera).

I added a flag to turn off this association between pitch and altitude.

I also added a function to directly update the camera with edge insets. Edge insets in this method are used to control the point on screen where the map rotates and where the center coordinate will be displayed.

Before | After
before | after

### CODE REVIEW

**Comment by julianrex** [comment]
@d-prukop is there any possible overlap with https://github.com/mapbox/mapbox-gl-native/issues/12203?

File: platform/ios/src/MGLMapView.h [hunk 1]
```diff
@@ -526,6 +526,13 @@ MGL_EXPORT IB_DESIGNABLE
  */
 @property(nonatomic, getter=isHapticFeedbackEnabled) BOOL hapticFeedbackEnabled;
 
+/**
+ A Boolean value that determines whether the updating pitch will also affect the altitude.
+ 
+ When this property is set to `YES`, pitch will work independently from altitude.
+ */
+@property(nonatomic, getter=isAltitudeUnaffectedByPitch) BOOL altitudeUnaffectedByPitch;
```
Comment (julianrex) [inline platform/ios/src/MGLMapView.h]: I think the opposite `altitudeAffectedByPitch` might be less confusing? Also, we should specify the default value.

**Comment by d-prukop** [comment]
It does seem like it has some overlap with #12203. @1ec5 do you want to talk about this approach?

File: platform/ios/src/MGLMapView.h [hunk 2]
```diff
@@ -922,6 +929,18 @@ MGL_EXPORT IB_DESIGNABLE
...
+ Moves the viewpoint to a different location without using a transition.
+ 
+ Transition animations can be defined by other classes and use this method to
+ update the map.
+ 
+ @param camera The new viewpoint.
+ @param padding The minimum padding (in screen points) that would be visible
+ 
+ */
+- (void)jumpToCamera:(MGLMapCamera *)camera edgePadding:(UIEdgeInsets)insets;
```
Comment (julianrex) [inline platform/ios/src/MGLMapView.h]: My understanding is that this essentially exposes core's `jumpTo` method. But it feels close to `-flyToCamera:withDuration:completionHandler:` with a 0 duration (which calls down to the internal method `-_flyToCamera:edgePadding:withDuration:peakAltitude:completionHandler:`.

I think this could be confusing. Thoughts?

Secondly, I think this method might need a completion handler.

File: platform/ios/src/MGLMapView.h (hunk 1 above)
Comment (d-prukop) [reply platform/ios/src/MGLMapView.h]: I chose this naming for the purpose of defaulting to false. I'm definitely open to better naming. When you get altitude from the internal GL core zoom level, pitch is taken into account in the calculation. Same thing when you set zoom level from altitude. I want just the pure altitude and the pure pitch. Any suggestions?

File: platform/ios/src/MGLMapView.h (hunk 2 above)
Comment (d-prukop) [reply platform/ios/src/MGLMapView.h]: I definitely need direct access to the edgePadding argument. I also don't want any animation since I'm controlling the animation outside of the iOS SDK. I looked into using one of the setCamera methods but I couldn't get it to work the way I wanted. Plus it appeared to drop frames vs this method. This could probably use a closer look.

File: platform/ios/src/MGLMapView.h (hunk 1 above)
Comment (fabian-guerra) [inline platform/ios/src/MGLMapView.h]: Is it OK if we specify that this property controls the camera updates and nothing more, something along the lines —feel free to add different wording—:

> A Boolean value that determines wether updating the camera pitch will also affect the altitude.
When this property is set to `YES`, pitch will work independently from altitude.
The default value of this property is `NO`.

Also I think the documentation needs a way to clarify the scope of this change and how people may take advantage, I know you already have a repo with experimental changes, but communicating how devs may take advantage of this I think is important.

In the SDK we follow the convention of avoid using negatives as part of boolean properties. As @julianrex said `isAltitudeAffectedByPitch` follows that pattern. But also is it more clear the intent using something like `isCameraAltitudeAffectedByPitch`?

File: platform/ios/src/MGLMapView.h (hunk 2 above)
Comment (fabian-guerra) [reply platform/ios/src/MGLMapView.h]: Right know we have two direct flavors to set the camera the `setCamera: XXX` and `flyToCamera`, `jumpToCamera` sounds like falls into the first flavor, would it be possible to add `setCamera:edgePadding:` instead?

File: platform/ios/src/MGLMapView.h [hunk 3]
```diff
@@ -529,12 +529,11 @@ MGL_EXPORT IB_DESIGNABLE
...
  
- When this property is set to `NO`, pitch will work independently from altitude.
- The default value of this property is YES.
- 
- Setting this property to NO will allow animation libraries outside of this SDK to control the
+ Setting this property to `NO` will allow animation libraries outside of this SDK to control the
  map camera without the built-in side effect that pitch influences altitude. This will make the
  results of using outside animation more predictable and easier to control.
+ 
+ The default value of this property is YES.
```
Comment (fabian-guerra) [inline platform/ios/src/MGLMapView.h]: We require all types be marked with \` to get rendered properly in our documentation. 
Nit: `YES`

**Comment by julianrex** [comment]
Thanks @d-prukop!

**Comment by friedbunny** [comment]
I’m excited about improvements to the camera system, these animations look quite nice. I have a couple questions about how we should implement these particular changes:

#### Should `MGLMapView.cameraAltitudeAffectedByPitch` be public API?
Or would it be acceptable to stash it in `MGLMapView_Private.h` and forward-declare in client implementations? As we prototype a new camera system, it feels like we’d be best served by not broadening/muddying the public API surface until we have a solid vision of what we want to support going forward.

#### Should pitch affect altitude?
Would it make sense to change the default behavior instead?

**Comment by d-prukop** [comment]
I'm OK with making this private since it will probably only be used by a new camera system. We should talk about the future of the camera!

I think the default behavior was put in there to mimic the way mapkit works. Whether we want that is a larger discussion.

File: platform/ios/src/MGLMapView.mm [hunk 4]
```diff
@@ -3374,7 +3382,8 @@ - (MGLMapCamera *)cameraForCameraOptions:(const mbgl::CameraOptions &)cameraOpti
     double zoomLevel = cameraOptions.zoom ? *cameraOptions.zoom : self.zoomLevel;
     CLLocationDirection direction = cameraOptions.angle ? mbgl::util::wrap(-MGLDegreesFromRadians(*cameraOptions.angle), 0., 360.) : self.direction;
     CGFloat pitch = cameraOptions.pitch ? MGLDegreesFromRadians(*cameraOptions.pitch) : _mbglMap->getPitch();
-    CLLocationDistance altitude = MGLAltitudeForZoomLevel(zoomLevel, pitch, centerCoordinate.latitude, self.frame.size);
+    CLLocationDegrees pitchForAltitude = self.isCameraAltitudeAffectedByPitch ? pitch : 0.0;
+    CLLocationDistance altitude = MGLAltitudeForZoomLevel(zoomLevel, pitchForAltitude, centerCoordinate.latitude, self.frame.size);
```
Comment (1ec5) [inline platform/ios/src/MGLMapView.mm]: Pitch is by definition affected by zoom level: when the map is tilted and the zoom level changes (whether programmatically or via a gesture), the viewing distance from the focal point and the viewpoint’s distance from the ground (the altitude) both change.

These changes affect the camera getters only. So the new property is a strange way of telling the getter to effectively lie about the camera relative to what mbgl thinks the camera is. Is there no way to derive the correct value from the existing return value, perhaps by exposing MGLAltitudeForZoomLevel publicly, as proposed in #5583?

File: platform/ios/src/MGLMapView.h [hunk 5]
```diff
@@ -526,6 +526,17 @@ MGL_EXPORT IB_DESIGNABLE
...
+/**
+ A Boolean value that determines whether the updating pitch will also affect the altitude.
+ 
+ Setting this property to `NO` will allow animation libraries outside of this SDK to control the
+ map camera without the built-in side effect that pitch influences altitude. This will make the
+ results of using outside animation more predictable and easier to control.
+ 
+ The default value of this property is `YES`.
+ */
+@property(nonatomic, getter=isCameraAltitudeAffectedByPitch) BOOL cameraAltitudeAffectedByPitch;
```
Comment (1ec5) [inline platform/ios/src/MGLMapView.h]: MGLMapView’s public API is expressed in terms of zoom levels, not in terms of altitude. Generally speaking, MGLMapView’s direct parameters are tied to 2D concepts for simplicity, while MGLMapCamera is better suited for expressing 3D concepts because it can account for two different frames of reference (viewpoint versus focal point). Mixing the two concepts would promote one frame of reference over the other, which would call into question the very purpose of having MGLMapCamera as a separate class.

As far as MGLMapView is concerned, pitch is a detail of MGLMapCamera. You’ll note that the rest of MGLMapView avoids referring to pitch, because normally the pitch is intertwined with other camera parameters and the camera’s frame of reference.

**Comment by julianrex** [comment]
Perhaps moving the property to `MGLCamera` is a solution here. The property name already hints that it might be in the wrong place. You'd need some additional scaffolding to pass it down appropriately though.

**Comment by d-prukop** [comment]
I'm not sure I can move it to MGLCamera. MGLMapView makes MGLCameras from mgbl::CameraOptions and I need to tell it how to handle those in the case where I want pitch and altitude separated.

File: platform/ios/src/MGLMapView.mm (hunk 4 above)
Comment (d-prukop) [reply platform/ios/src/MGLMapView.mm]: I don't think mbgl comes in to this at all. As far as I can tell this is all happening in MGLGeometry in the iOS SDK. The pure values in mbgl are what I want.

Comment (d-prukop) [reply platform/ios/src/MGLMapView.mm]: > The pure values in mbgl are what I want.

This is wrong.
The problem with exposing MGLAltitudeForZoomLevel is that, if you use setCamera, mbgl sets it's state from the result of MGLZoomLevelForAltitude. The stored zoomLevel is affected by pitch. I want to prevent the mbgl state from getting that value and storing the zoomLevel without regards to pitch.

File: platform/ios/src/MGLMapView.h (hunk 5 above)
Comment (d-prukop) [reply platform/ios/src/MGLMapView.h]: Right now if I want to change pitch I have to use MGLMapCamera. When the MGLMapCamera is passed to cameraOptionsObjectForAnimatingToCamera, mbgl updates it's state with the zoomLevel calculated from MGLZoomLevelForAltitude.

All of the setCamera functions use cameraOptionsObjectForAnimatingToCamera.

Comment (1ec5) [reply platform/ios/src/MGLMapView.h]: My point is that MGLMapView’s API shouldn’t even mention pitch. If mentions pitch, it should provide the full complement of ways to interact with the pitch, which it doesn’t by design.

Think of camera as a set of parameters that get updated together. They have to be updated together in order to maintain consistency. If you want to update the pitch without updating other parameters, you set the other parameters to special values. For example, set heading to -1 or centerCoordinate to an invalid coordinate.

File: platform/ios/src/MGLMapView.mm (hunk 4 above)
Comment (1ec5) [reply platform/ios/src/MGLMapView.mm]: If you want to change the zoom level without changing the pitch, set a camera whose pitch is -1.

Comment (1ec5) [reply platform/ios/src/MGLMapView.mm]: @d-prukop and I chatted about this today. The pitch is considered twice when calculating the camera options, once for setting the camera’s actual pitch and once in calculating the zoom level. I would’ve thought that this is ideal, because the zoom level on a tilted map is really the viewing distance, but I suppose if the intention is to iteratively change the viewing distance by getting and setting the camera iteratively, then there could be a sort of feedback effect.

The essence of my concern here is that we’re adding a separate code path for a specific use case. If we really believe that the altitude-to-zoom-level conversion shouldn’t be affected by the pitch, then why not remove the pitch argument from `MGLZoomLevelForAltitude()` and `MGLAltitudeForZoomLevel()` and always assume a pitch of 0°? The obvious consequence would be that the MGLMapCamera returned by the `camera` getter would have the same `altitude` regardless of the pitch.

In other words, `isCameraAltitudeAffectedByPitch` changes the meaning of `altitude` from “shortest distance between the viewpoint and the ground” to “viewing distance from the focal point”, essentially a linear, physical-unit version of `zoomLevel`. Maybe that’s what the use case above calls for: a property for explicitly setting the zoom level of a camera instead of the altitude. We already have that in MGLMapSnapshotOptions, so why not MGLMapCamera? One reason why is that it’s awkward for an object to sport mutually exclusive properties. Any future attempt at making the `isCameraAltitudeAffectedByPitch` option public or defaulting it to `YES` will need to consider whether we instead need a `zoomLevel` property on MGLMapCamera.

File: platform/ios/src/MGLMapView.h [hunk 6]
```diff
@@ -856,6 +856,15 @@ MGL_EXPORT IB_DESIGNABLE
...
 - (void)setCamera:(MGLMapCamera *)camera animated:(BOOL)animated;
 
+/**
+ Moves the viewpoint to a different location without using a transition.
+ 
+ @param camera The new viewpoint.
+ @param edgePadding The minimum padding (in screen points) that would be visible
+ 
+ */
+- (void)setCamera:(MGLMapCamera *)camera edgePadding:(UIEdgeInsets)edgePadding;
```
Comment (1ec5) [inline platform/ios/src/MGLMapView.h]: This method should accept an `animated` parameter. Just as there’s a `-setVisibleCoordinateBounds:animated:` and `-setVisibleCoordinateBounds:edgePadding:animated:`, there should be `-setCamera:animated:` and `-setCamera:edgePadding:animated:`. It seems unfair to force the developer to choose between specifying an edge padding and using the built-in animation capabilities, even if your particular use case doesn’t require those capabilities.

File: platform/ios/src/MGLMapView.mm [hunk 7]
```diff
@@ -240,6 +240,9 @@ @interface MGLMapView () <UIGestureRecognizerDelegate,
 @property (nonatomic) MGLUserLocation *userLocation;
 @property (nonatomic) NSMutableDictionary<NSString *, NSMutableArray<MGLAnnotationView *> *> *annotationViewReuseQueueByIdentifier;
 
+/// A Boolean value that determines whether the updating pitch will also affect the altitude.
+@property(nonatomic, getter=isCameraAltitudeAffectedByPitch) BOOL cameraAltitudeAffectedByPitch;
```
Comment (1ec5) [inline platform/ios/src/MGLMapView.mm]: A more natural name for the getter would be `cameraAltitudeIsAffectedByPitch` or, given https://github.com/mapbox/mapbox-gl-native/pull/12518#discussion_r207418088, `cameraAltitudeIsViewingDistance`.

**Comment by d-prukop** [comment]
Abandoning this PR. Going to start a new one
